extensions/elasticsearch/elasticsearch-7/src/main/java/com/hazelcast/jet/elastic/ElasticSourceBuilder.java

```diff
@@ -0,0 +1,236 @@
+     */
+    @Nonnull
+    public ElasticSourceBuilder<T> clientFn(@Nonnull SupplierEx<RestClientBuilder> clientFn) {
+        this.clientFn = checkNonNullAndSerializable(clientFn, "clientFn");
+        return this;
+    }
+
+    /**
+     * Set the search request supplier function
+     * <p>
+     * The connector executes this search request to retrieve documents from Elasticsearch.
+     * <p>
+     * For example, to create SearchRequest limited to an index `logs`:
+     * <pre>{@code
+     * builder.searchRequestFn(() -> new SearchRequest("logs"))
+     * }</pre>
+     *
+     * This parameter is required.
+     *
+     * @param searchRequestFn search request supplier function
+     */
+    @Nonnull
+    public ElasticSourceBuilder<T> searchRequestFn(@Nonnull SupplierEx<SearchRequest> searchRequestFn) {
+        this.searchRequestFn = checkSerializable(searchRequestFn, "searchRequestFn");
+        return this;
+    }
+
+    /**
+     * Set the function to map SearchHit to a pipeline item
+     * <p>
+     * For example, to map a SearchHit to a value of a field `productId`:
+     * <pre>{@code
+     * builder.mapToItemFn(hit -> (String) hit.getSourceAsMap().get("productId"))
+     * }</pre>
+     *
+     * This parameter is required.
+     *
+     * @param mapToItemFn maps search hits to output items
+     */
+    @Nonnull
+    @SuppressWarnings("unchecked")
+    public <T_NEW> ElasticSourceBuilder<T_NEW> mapToItemFn(@Nonnull FunctionEx<? super SearchHit, T_NEW> mapToItemFn) {
+        ElasticSourceBuilder<T_NEW> newThis = (ElasticSourceBuilder<T_NEW>) this;
+        newThis.mapToItemFn = checkSerializable(mapToItemFn, "mapToItemFn");
+        return newThis;
+    }
+
+    /**
+     * Set the function that provides {@link RequestOptions}
+     * <p>
+     * It can either return a constant value or a value based on provided request.
+     * <p>
+     * For example, use this to provide a custom authentication header:
+     * <pre>{@code
+     * sourceBuilder.optionsFn((request) -> {
+     *     RequestOptions.Builder builder = RequestOptions.DEFAULT.toBuilder();
+     *     builder.addHeader("Authorization", "Bearer " + TOKEN);
+     *     return builder.build();
+     * })
+     * }</pre>
+     *
+     * @param optionsFn function that provides {@link RequestOptions}
+     * @see <a
+     * href="https://www.elastic.co/guide/en/elasticsearch/client/java-rest/current/java-rest-low-usage-requests.html">
+     * RequestOptions in Elastic documentation</a>
+     */
+    @Nonnull
+    public ElasticSourceBuilder<T> optionsFn(@Nonnull FunctionEx<? super ActionRequest, RequestOptions> optionsFn) {
+        this.optionsFn = checkSerializable(optionsFn, "optionsFn");
+        return this;
+    }
+
+    /**
+     * Enable slicing
+     * <p>
+     * Number of slices is equal to globalParallelism (localParallelism * numberOfNodes)
+     * <p>
+     * Use this option to read from multiple shards in parallel.
```

Comment:
I think we should enhance the JavaDoc here since slicing can be done on a single shard too.
Additionally, when slicing is used with co-located reading I guess the number of slices change ?